### PR TITLE
Update GHA workflow to build all Docker images

### PR DESCRIPTION
**Pull request by stumpylog**
## Proposed change

This PR brings all the images and workflow steps from the builder repo back into the main repo, so we no longer need to manage certain Python dependencies in two locations.  It's generally less confusing to new contributors, simpler to reason about versions, and easy to manage new packages, etc.

95% of these files are copied over from the builder repo, with the minimal changes to incorporate here.

### Features
- Python dependency versions parsed from Pipfile.lock
- dependabot will keep Pipfile.lock updated for us
- Other versions set in `get-build-json.py`
- Front end compilation is now separate job
- Front end compilation only happens if the frontend changed OR the frontend hasn't been built yet for this Git ref
- Reusable workflows are used to encapsulate frontend vs backend jobs.
- Reusable workflow used for building our intermediate images
- Our primary Dockerfile is now easier to understand where to put new packages needed for building vs runtime
- Concurrency groups keep us from building the same intermediate image multiple times (jobs will pause though, instead of cancelling in progress work)

### Downsides

- Our workflow will take longer to run the first time one of the certain dependencies is updated
- The workflow looks a lot longer now
- If there's some issue with a built/tagged image, it would need to be manually deleted and re-run

### Open Questions
- I'm not sure pull requests from forked builds will be able to add new intermediate image if needed
- Do we have any concerns about pull requests and allowing them to push to our image registry?
- I named the intermediate  images something, anyone have other name ideas?
- Does the frontend building differ between architectures or could it be done only on amd64? 

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other (please explain)

## Checklist:

- [x] I have read & agree with the [contributing guidelines](https://github.com/paperless-ngx/paperless-ngx/blob/main/CONTRIBUTING.md).
- [ ] If applicable, I have tested my code for new features & regressions on both mobile & desktop devices, using the latest version of major browsers.
- [ ] If applicable, I have checked that all tests pass, see [documentation](https://paperless-ngx.readthedocs.io/en/latest/extending.html#back-end-development).
- [x] I have run all `pre-commit` hooks, see [documentation](https://paperless-ngx.readthedocs.io/en/latest/extending.html#code-formatting-with-pre-commit-hooks).
- [ ] I have made corresponding changes to the documentation as needed.
- [x] I have checked my modifications for any breaking changes.

### CODE REVIEW

**Comment by stumpylog** [comment]
The naming question is about the intermediate names.  So `paperless-ngx/paperless-ngx/builder/qpdf:10.6.3`, for example.  I copied it from the builder, but it could be named something else if there's a strong feeling in the group for a different name

File: .github/workflows/reusable-ci-backend.yml
Comment (shamoon) [inline .github/workflows/reusable-ci-backend.yml:1]: Curious, why not name this just `ci-backend.yml`?

Comment (stumpylog) [reply .github/workflows/reusable-ci-backend.yml:1]: The file?  I wanted to name to be clear it wasn't triggered by some event, only by another workflow.

**Comment by shamoon** [comment]
> The naming question is about the intermediate names. So `paperless-ngx/paperless-ngx/builder/qpdf:10.6.3`, for example. 

Oh, I see no reason to change that, consider it an homage to the `builder` #rip

> The frontend gets built for each arch currently, and the final image copies from the matching arch, so if there are differences, they'll be accounted for if we leave it alone.

Perhaps that’s safest I just can’t imagine why there would be differences. Maybe I’ll pull a couple and compare a bit. I assume it would be faster not to do each though?

File: .github/workflows/reusable-ci-backend.yml
Comment (shamoon) [reply .github/workflows/reusable-ci-backend.yml:1]: Yea the file. Ok, I see. And maybe that’s a convention

**Comment by stumpylog** [comment]
> I assume it would be faster not to do each though?

It would be, but with the changes I made to only build the frontend once per branch (then again if src-ui changes), the speedup would only affect a first run or runs with frontend changes.

**Comment by shamoon** [comment]
Again, very cool. I appreciate that its more modular, hopefully should help with sniffing out issues if they arise. And if I understand it correctly theres no more worrying about a base image version (same idea still happens, just doesnt need to be manually tagged, etc. any more, yea?) 

A couple thoughts:

1. Should we wait until after 1.7.0 is released (assuming thats pretty close) before merging this? After the release we can let this loose on `dev` and have some time before the next release to test it.
2. In general Im not sure how much genuine testing folks are gonna do of this (but maybe Im wrong about that?). As you (and schnuffle!) are clearly the guru of this, I think we all trust you're happy with where its at and so I would be happy to merge this (again, timing as per above) as I dont really see an obvious errors or anything like that. Of course curious how others feel about that.

**Comment by stumpylog** [comment]
- No more versioning between the repos, it'll always be self contained to a branch or tag
- I targets this to `dev` so it wouldn't affect the 1.7.0 release.  It's too big a change to think of trying for that, which seems to be getting close to complete
- Once the beta is completed, I'd be fine fixing any conflicts and merging this.  I've run the resulting image and it started/consumed generally, so it seems alright in the output product.

**Comment by qcasey** [comment]
This is wonderful!

We should be able to push to `dev` during beta testing, so merging this before it's released is fine IMO.

> I'm not sure pull requests from forked builds will be able to add new intermediate image if needed

> Do we have any concerns about pull requests and allowing them to push to our image registry?

IIRC, only members of the organization are allowed to push. This is ideal for security. #106 

> I named the intermediate images something, anyone have other name ideas?

If you mean the [builder/psycopg2](https://github.com/orgs/paperless-ngx/packages/container/package/paperless-ngx%2Fbuilder%2Fpsycopg2) packages, I think this name is fine.

It does seem to build correctly. I'm able to run `ghcr.io/paperless-ngx/paperless-ngx:feature-build-all-images` without any issues. Stellar job here.

**Comment by stumpylog** [comment]
There are some merge conflicts with `beta`, but not as bad as I thought.  So I'm good to merge it pretty much anytime after people have had a look.

**Comment by stumpylog** [comment]
Just while I had some time:
- non-Python versions are now configured in a single location
- Because the Dockerfile is getting specialized to the workflow, added a script to help locally building any of the images
- I checked the frontend files, they don't differ by architecture, so now that image builds only for amd64
- Some small improvements for Docker build caching on the frontend image

**Comment by stumpylog** [comment]
The above mentioned merge conflicts are resolved and this is ready now.  I tested with it creating new packages on my fork and it ran all the way.  I've also built the images locally using the util script.

**Review by shamoon — APPROVED** [review]
Let’s boogie

**Pull request merged**
